### ISSUE TITLE
Watch Caddyfile for changes

### ISSUE THREAD

**Issue by AndreKR** [caddyserver/caddy#1806]
I would like to have Caddy watch its Caddyfile and automatically reload (like SIGUSR1) if it changes.

I can send a PR, but before doing that I would like some guidance on a few thoughts and questions:
* This feels like it's a configuration option for the `defaultLoader`, correct? Caddyfile Loaders should be configured by environment variables and command line options, correct?
* This should probably be done in the core instead of a plugin, correct? There are a few ways to do this as a plugin with minimal core changes, but they all feel like workarounds. (Like: parsing the Caddfile a second time to find out about all the imports or providing a new. Or: Providing a new hook that is called when a list of all loaded files is available - which would make the plugin half Caddyfile Loader and half Plugin.)
* Name of the flag/variable? `auto-reload`/`CADDY_AUTO_RELOAD`? `watch-caddyfile`/`CADDY_WATCH_CADDYFILE`?

**Comment by mholt** [comment]
As I think about this, I think it should be a separate loader plugin. Could be invoked with a command line flag like `-watch-conf-changes` or something like that.

So, while I think this is a very valid request, and could be a valuable feature, I do think this would be better fit as a plugin, so I will close the issue. But I'm sure quite a few people would use a plugin like this!

**Comment by AndreKR** [comment]
Here's an early implementation as a plugin, please have a look:
https://github.com/AndreKR/caddyfile-watch

`dispenser.go` and `lexer.go` are verbatim copies from Caddy, `parse.go` has a [tiny modification](https://github.com/AndreKR/caddyfile-watch/blob/fe6c1927472488a07edf65c79eeaad8ffdf5a0e1/parse.go#L246) and `run.go` copies two functions. This way patches to Caddy can easily be applied to the plugin (although I would still prefer to get the imported files from the core and avoid all that duplicate code).

We're going to need an API in Caddy to trigger a reload of the instance. At the moment I trigger the reload by SIGUSR1 on Posix and show an error message on Windows.

**Comment by mholt** [comment]
I'm confused why you need to parse the Caddyfile yourself, etc? Don't you just load it from disk and then, if it changes, send the signal to reload?

As far as actually telling Caddy to reload, I know it's a bit clunky right now. Will improve on that, might not even need the API to do that...

**Comment by mholt** [comment]
Oh, you parse it so you can also watch all the imported files for changes? (Is that the only reason? If so, we may be able to accommodate...)

**Comment by AndreKR** [comment]
> why you need to parse the Caddyfile yourself
Yes, just to find out about all involved import files.

> telling Caddy to reload
I would think that it's just a matter of wrapping [this](https://github.com/mholt/caddy/blob/bd67ec99f01797931a0787b9c2de0f04fa0e5980/sigtrap_posix.go#L48) in a function and exposing it to the plugin?

**Comment by mholt** [comment]
@AndreKR I wonder if we can write an exported function to report all the files that are, or will be, imported... 🤔 Maybe a method on CaddyfileInput or something.

As for telling Caddy to reload, we could wrap that in a function, or we could give Caddyfile loaders access to the []*Instance -- or the single *Instance, I guess, for when Caddy as a command which is most of the time -- on which you can call Restart().

**Comment by AndreKR** [comment]
It doesn't really fit the Caddyfile loader plugin API, because it would have to run *after* the confLoader/defaultLoader has done its work. That's what I meant in my first post.

**Comment by mholt** [comment]
Gonna re-open to see how we can handle this.

**Comment by zaolin** [comment]
Maybe https://github.com/fsnotify/fsnotify could handle this. :)

**Comment by AndreKR** [comment]
@zaolin Yes, that's what I'm using to get the filesystem events of course.

**Comment by maciekish** [comment]
Any news on this please? Im running Caddy in a VM with the config folder exposed via SMB. It would simplify config changes a lot (no more SSH to reload config).

**Comment by lbguilherme** [comment]
I currently use the following script to start Caddy, as workaround:

```bash
function check {
    HASH=$(find /data/caddy -type f -print0 | sort -z | xargs -0 sha256sum | sha256sum || echo 0)
    while [ true ]; do
        sleep 2
        NEW_HASH=$(find /data/caddy -type f -print0 | sort -z | xargs -0 sha256sum | sha256sum || echo 0)
        if [ "$HASH" != "$NEW_HASH" ]; then
            HASH=$NEW_HASH
            echo "~~~SENDING RELOAD SIGNAL~~~"
            kill -s USR1 $(cat /tmp/caddy.pid)
        fi
    done
}

check &

exec caddy -agree -email aaa@example.com -log stdout -conf /data/caddy/Caddyfile -grace 60s -pidfile /tmp/caddy.pid
```

**Comment by awoodbeck** [comment]
Please see the PR with my approach.  I'm relying on an event call that emits _after_ the Caddyfile has been completely processed.  The corresponding Caddyfile plug-in doesn't return `Input`, but rather hooks into the event.  I'm not sure if this is the best approach, but it was the lightest touch I could make without modifying any more of the API.  I would appreciate it if someone with more knowledge of the code base would weigh in on this approach.

**Comment by mholt** [comment]
Implemented in Caddy 2 with 6e2fabb as the `--watch` flag. Please try it out!
